### PR TITLE
Commissions by partner

### PR DESCRIPTION
**Pull request by javierjcf**
We can set a diferent commision for the same agent in each partner.

### CODE REVIEW

File: sale_commission/i18n/es.po [hunk 1]
```diff
@@ -29,13 +26,20 @@ msgid "(keep empty for making the settlement of all agents)"
 msgstr "(mantener vacío para realizar la liquidación de todos los agentes)"
 
 #. module: sale_commission
+#: code:addons/sale_commission/models/partner_agent.py:54
+#, python-format
+msgid "A commission has been assigned by sections that does not match that defined for the agent by default, so that these sections shall apply only on this bill."
+msgstr "A commission has been assigned by sections that does not match that defined for the agent by default, so that these sections shall apply only on this bill."
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate this

File: sale_commission/i18n/es.po [hunk 2]
```diff
@@ -74,11 +81,22 @@ msgid "Agents"
...
 msgstr "Agentes y comisiones"
 
 #. module: sale_commission
+#: view:res.partner:sale_commission.view_partner_form_agent
+msgid "Agents / Commissions"
+msgstr "Agentes / Comisiones"
+
+#. module: sale_commission
+#: view:res.partner:sale_commission.view_partner_form_agent
+msgid "Agents and Commissions"
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Why don't you use the same spelling in all places (`Agents & commissions`) and thus there's no similar strings to translate?

File: sale_commission/i18n/es.po [hunk 3]
```diff
@@ -182,6 +206,11 @@ msgid "Commissions"
 msgstr "Comisiones"
 
 #. module: sale_commission
+#: field:res.partner.agent,company_id:0
+msgid "Company"
+msgstr "Company"
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate this

File: sale_commission/i18n/es.po [hunk 4]
```diff
@@ -254,19 +285,34 @@ msgid "Date month"
...
+#: field:res.partner.agent,display_name:0
+#: field:sale.commission,display_name:0
+#: field:sale.commission.make.invoice,display_name:0
+#: field:sale.commission.make.settle,display_name:0
+#: field:sale.commission.section,display_name:0
+#: field:sale.commission.settlement,display_name:0
+#: field:sale.commission.settlement.line,display_name:0
+#: field:sale.order.line.agent,display_name:0
+msgid "Display Name"
+msgstr "Display Name"
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate

File: sale_commission/i18n/es.po [hunk 5]
```diff
@@ -254,19 +285,34 @@ msgid "Date month"
...
 msgstr "Agente externo"
 
 #. module: sale_commission
-#: selection:sale.commission,commission_type:0 field:sale.commission,fix_qty:0
-msgid "Fixed percentage"
-msgstr "Porcentaje fijo"
+#: code:addons/sale_commission/models/partner_agent.py:53
+#, python-format
+msgid "Fee installments!"
+msgstr "Fee installments!"
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate

File: sale_commission/i18n/es.po [hunk 6]
```diff
@@ -375,7 +425,21 @@ msgid "Keep empty to use the current date"
...
+#: field:res.partner.agent,__last_update:0
+#: field:sale.commission,__last_update:0
+#: field:sale.commission.make.invoice,__last_update:0
+#: field:sale.commission.make.settle,__last_update:0
+#: field:sale.commission.section,__last_update:0
+#: field:sale.commission.settlement,__last_update:0
+#: field:sale.commission.settlement.line,__last_update:0
+#: field:sale.order.line.agent,__last_update:0
+msgid "Last Modified on"
+msgstr "Last Modified on"
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate

File: sale_commission/i18n/es.po [hunk 7]
```diff
@@ -477,19 +554,29 @@ msgid "Refund Journal"
 msgstr "Diario de Reembolso"
 
 #. module: sale_commission
+#: field:account.invoice,picking_ids:0
+msgid "Related Pickings"
+msgstr "Related Pickings"
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate

File: sale_commission/i18n/es.po [hunk 8]
```diff
@@ -477,19 +554,29 @@ msgid "Refund Journal"
...
+
+#. module: sale_commission
+#: help:account.invoice,picking_ids:0
+msgid "Related pickings (only when the invoice has been generated from the picking)."
+msgstr "Related pickings (only when the invoice has been generated from the picking)."
```
Comment (pedrobaeza) [inline sale_commission/i18n/es.po]: Translate

File: sale_commission/models/res_partner.py [hunk 9]
```diff
@@ -36,6 +36,13 @@ class ResPartner(models.Model):
     settlements = fields.One2many(
         comodel_name="sale.commission.settlement", inverse_name="agent",
         readonly=True)
+    commission_ids = fields.One2many(comodel_name='res.partner.agent',
+                                     inverse_name='partner_id',
+                                     string='Agents')
+    default_commission = fields.Boolean(
```
Comment (pedrobaeza) [inline sale_commission/models/res_partner.py]: This field should be in res.partner.agent level, not here. The idea is:
- You set the agents at customer (partner) level
- You check for each agent if they used the default commission set as agent level
- Or you specify a specific one for this customer

**Comment by javierjcf** [comment]
Ok,
I underestand if i do it this way, i must remove the agents field (because it make not sense have two diferent fields to set the agents in the partner) and put a post_init_hook to migrate the data from the agents fields.
Is that ok?

**Comment by pedrobaeza** [comment]
Yeah, that's it, @javierjcf!

**Comment by javierjcf** [comment]
Done.
I prefer a post-migration script instead of post_init_hook, because we only want do the script when the version change, and not after any installation.

File: sale_commission/__openerp__.py [hunk 10]
```diff
@@ -6,7 +6,7 @@
 
 {
     'name': 'Sales commissions',
-    'version': '8.0.2.4.0',
+    'version': '8.0.2.4.1',
```
Comment (pedrobaeza) [inline sale_commission/__openerp__.py]: It should be 8.0.3.0.0 as it includes a major change with migration script (https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers)

File: sale_commission/migrations/8.0.2.4.1/post-migration.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+__name__ = ('Copy agents to the new One2many commission_ids')
+
+
+def migrate_agents_field(cr):
+    cr.execute("""
+        INSERT INTO res_partner_agent
+        (partner_id, agent_id, commission_id, default_commission, company_id)
+        (select par.partner_id, agent_id, rp.commission, True, null
```
Comment (pedrobaeza) [inline sale_commission/migrations/8.0.2.4.1/post-migration.py:13]: Instead null, put the partner company.

File: sale_commission/migrations/8.0.2.4.1/post-migration.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        FROM partner_agent_rel par
```
Comment (pedrobaeza) [inline sale_commission/migrations/8.0.2.4.1/post-migration.py:14]: Isn't this table dropped when updating the module? If so, you have to rename it on the pre-migration script.

File: sale_commission/models/partner_agent.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+# © 2015 Comunitea Servicios Tecnológicos (<http://www.comunitea.com>)
```
Comment (pedrobaeza) [inline sale_commission/models/partner_agent.py:2]: Don't add 2 copyright lines (and it's 2016 I suppose :wink: )

File: sale_commission/models/partner_agent.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
```
Comment (pedrobaeza) [inline sale_commission/models/partner_agent.py:1]: The file should be `res_partner_agent.py`

**Comment by pedrobaeza** [comment]
Thanks for the efforts you are making, Javier. Let's go for the last steps:
- You are making the commit messages in Spanish, but it should be in English. I suggest you squash the commit history and write a new global message in English.
- The coverage is dropping. Can you add tests for the non-covered parts for keeping it?

**Comment by javierjcf** [comment]
One line is difficult to cover (and i think  it was no covered before the pull request):
 if lines and lines[0]._name != 'account.invoice.line':
            return res
I think is not neccesarry this lines but i prefer don't remove it.

**Comment by javierjcf** [comment]
Hi Pedro, 
I check there is no need to pre_init hook, the table partner_agent_rel is not removed.
I check it updating the module from OCA 8.0 branch to my branch, in a clean database with demo data.
I see that the data was migrated and the many2many table already exists.

**Comment by DarkoNikolovski** [comment]
Hi All,

Is there any chance that this PR will be merged soon ?

**Comment by pedrobaeza** [comment]
There are conflicts, so you must rebase

**Comment by DarkoNikolovski** [comment]
Rebased locally. May I push the changes to 8.0 ?

**Comment by pedrobaeza** [comment]
You cannot push. It should be the author. Another option is to close this PR and you make another one with your rebase, superseeding this one.

**Comment by javierjcf** [comment]
Hi, 
I did rebase and push.

**Comment by DarkoNikolovski** [comment]
Thanks.

**Comment by DarkoNikolovski** [comment]
Any news when this PR would be merged ?

**Comment by pedrobaeza** [comment]
@DarkoNikolovski, you haven't even reviewed the PR, so please do it.

**Comment by DarkoNikolovski** [comment]
Sorry that I didn't comment here but I installed the module locally and tested it before two weeks. It was everything OK.

**Comment by yajo** [comment]
Hi, please make bots go :white_check_mark: and merge will come faster :wink:

**Comment by javierjcf** [comment]
Hi, i did tests for the uncover code of my PR.

**Comment by DarkoNikolovski** [comment]
Hi. I see all checks are passed . May we merge it now ?

File: sale_commission/security/ir.model.access.csv [hunk 15]
```diff
@@ -1,5 +1,6 @@
 "id","name","model_id:id","group_id:id","perm_read","perm_write","perm_create","perm_unlink"
 "access_sale_commission_manager","access_sale_commission_manager","model_sale_commission","base.group_sale_manager",1,1,1,1
+"access_res_partner_agent","access_res_partner_agent","model_res_partner_agent","base.group_sale_manager",1,1,1,1
```
Comment (atchuthan) [inline sale_commission/security/ir.model.access.csv:3]: Needs minimum READ access in base.group_user for agent_lines_company as it is added to res.partner using One2many field

**Review by atchuthan — APPROVED** [review]
Code review :+1: 
minor changes pointed out in security access and please test it using users other than Admin too.

**Comment by Derter** [comment]
Can we expect a merge here soon?

**Comment by pedrobaeza** [comment]
There are still comments unattended and 2 approvals are needed (one from one of the repo commiters).

**Comment by Derter** [comment]
Hello Guys,
did this PR failed now? Does it need new review? 

Whats the next step that needs to happen before we can merge?

**Comment by yajo** [comment]
Not sure why runbot's ❌... any clues?

**Comment by pedrobaeza** [comment]
```
2017-02-15 12:44:35,240 187 WARNING openerp_test openerp.models: method res.partner._check_employee: @constrains parameter 'employee' is not writeable
```

**Comment by Derter** [comment]
Would appreciate a merge of this one! Thank you.

**Comment by yajo** [comment]
I'm afraid we can't merge with a red runbot... You could try rebasing to see if it fixes it.

File: sale_commission/models/__init__.py [hunk 16]
```diff
@@ -8,3 +8,4 @@
 from . import sale_order
 from . import account_invoice
 from . import settlement
+from . import res_partner_agent
```
Comment (Daniel-CA) [inline sale_commission/models/__init__.py:11]: Add end line

File: sale_commission/views/sale_order_view.xml [hunk 17]
```diff
@@ -13,13 +13,15 @@
                     <field name="commission_free"/>
                     <field name="agents"
                            attrs="{'readonly': [('commission_free', '=', True)]}"
+                           context="{'partner_id': partner_id}"
```
Comment (scigghia) [inline sale_commission/views/sale_order_view.xml:16]: this line, give me an error opening a draft sale order.

File: sale_commission/views/sale_order_view.xml [hunk 18]
```diff
@@ -13,13 +13,15 @@
...
                            widget="one2many_tags"
                     />
                 </xpath>
                 <xpath expr="//field[@name='order_line']/form//field[@name='address_allotment_id']" position="after">
                     <field name="commission_free"/>
                     <field name="agents"
-                           attrs="{'readonly': [('commission_free', '=', True)]}"/>
+                           attrs="{'readonly': [('commission_free', '=', True)]}"
+                           context="{'partner_id': partner_id}"/>
```
Comment (scigghia) [inline sale_commission/views/sale_order_view.xml:24]: same as above

**Comment by pedrobaeza** [comment]
Are you still interested in this? CIs are red, there are conflicting files. Note that there's an alternative in #115

**Comment by pedrobaeza** [comment]
Closing as no answer.

**Pull request closed without merging**